runtime/vm/CRIUHelpers.cpp

```diff
@@ -389,4 +402,70 @@ runInternalJVMRestoreHooks(J9VMThread *currentThread)
 	return result;
 }
 
+BOOLEAN
+runDelayedIdentityOperations(J9VMThread *currentThread)
+{
+	pool_state walkState = {0};
+	J9JavaVM *vm = currentThread->javaVM;
+	J9Pool *delayedSingleThreadModeOpRecords = vm->checkpointState.delayedSingleThreadModeOpRecord;
+	J9InternalVMFunctions* vmFuncs = vm->internalVMFunctions;
+	J9DelayedSingleThreadModeOpRecord *delayedSingleThreadModeOp = (J9DelayedSingleThreadModeOpRecord*) pool_startDo(delayedSingleThreadModeOpRecords, &walkState);
+	omrthread_monitor_t monitorPtr = NULL;
+	BOOLEAN rc = TRUE;
+
+	while (NULL != delayedSingleThreadModeOp) {
+		j9object_t instance = J9_JNI_UNWRAP_REFERENCE(delayedSingleThreadModeOp->globalObjectRef);
+		if (!VM_ObjectMonitor::inlineFastObjectMonitorEnter(currentThread, instance)) {
+			rc = objectMonitorEnterNonBlocking(currentThread, instance);
+			if (J9_OBJECT_MONITOR_BLOCKING == rc) {
+				/* owned by another thread */
+				rc = FALSE;
+				goto done;
+			} else if (J9_OBJECT_MONITOR_ENTER_FAILED(rc)) {
+				/* not possible if the the application was able to call notify earlier */
+				Assert_VM_unreachable();
+			}
+		}
+		if (!VM_ObjectMonitor::getMonitorForNotify(currentThread, instance, &monitorPtr, true)) {
+			if (NULL != monitorPtr) {
+				/* another thread owns the lock, shouldn't be possible */
+				Assert_VM_unreachable();
+			} else {
+				/* no waiters */
+				goto next;
+			}
+		}
+
+		switch(delayedSingleThreadModeOp->operation) {
+		case J9_SINGLE_THREAD_MODE_OP_NOTIFY:
+			rc = 0 == omrthread_monitor_notify(monitorPtr);
+			break;
+		case J9_SINGLE_THREAD_MODE_OP_NOTIFY_ALL:
+			rc = 0 == omrthread_monitor_notify_all(monitorPtr);
+			break;
+		default:
+			Assert_VM_unreachable();
+			break;
+		}
+
+		if (!rc) {
+			goto done;
+		}
+
+		if (!VM_ObjectMonitor::inlineFastObjectMonitorExit(currentThread, instance)) {
+			if (0 != objectMonitorExit(currentThread, instance)) {
+				Assert_VM_unreachable();
+			}
+		}
+
+		monitorPtr = NULL;
+		vmFuncs->j9jni_deleteGlobalRef((JNIEnv*) currentThread, delayedSingleThreadModeOp->globalObjectRef, JNI_FALSE);
```

Comment:
Each item should be removed from the pool after the its been notified - use `pool_removeElement` which says it is safe to use during an iteration